compiler/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -500,7 +500,13 @@ impl<'interner> Monomorphizer<'interner> {
             },
             HirExpression::Literal(HirLiteral::Unit) => ast::Expression::Block(vec![]),
             HirExpression::Block(block) => self.block(block.statements)?,
-            HirExpression::Unsafe(block) => self.block(block.statements)?,
+            HirExpression::Unsafe(block) => {
+                let was_in_unconstrained_function = self.in_unconstrained_function;
+                self.in_unconstrained_function = true;
+                let res = self.block(block.statements);
+                self.in_unconstrained_function = was_in_unconstrained_function;
```

Comment:
This seems like we're treating the contents of an `unsafe` block as unconstrained. This isn't the case however as it's still constrained but the only difference is that we can call into unconstrained functions.